### ISSUE TITLE
Interaction with OverloadedRecordFields

### ISSUE THREAD

**Issue by danidiaz** [well-typed/generics-sop#18]
One thing I like of this library is the ability to wrap every field with a type constructor, usually some Applicative. 

In plain records (let's not consider sum types) it would be nice if we could manipulate the corresponding representation using the same "labels" that we use for the original record, allowing us to get and set individual "wrapped" values.  Perhaps the upcoming [OverloadedRecordFields](https://ghc.haskell.org/trac/ghc/wiki/Records/OverloadedRecordFields) extension could help with this?

Consider a type like

`data Wrapped f r = ...`

with a constructor like 

`makeWrapped :: (Generic r, Code r ~ '[ xs ]) => NP f xs -> Wrapped f r `

With OverloadedRecordFields ([here ](http://hackage.haskell.org/package/overloaded-records)is an implementation package based in the proposal) perhaps we could define an instance like

`instance (HasField symbol r a) => HasField symbol (Wrapped f r) (f a) where ...`

Would this make sense? Or perhaps the idea of directly accessing individual fields of the representation is misguided?

**Comment by kosmikus** [comment]
If I understand this correctly, then yes, it would certainly make sense. In order to define an instance such as you suggest generically, though, we'd have to have access to the field names on the type level, and in `generics-sop`, we currently do not. I'm working on type-level metadata for `generics-sop` based on the type-level metadata provided for GHC generics in 8.0.1, but there are still some difficulties (and bugs in GHC) to bridge. It also doesn't seem like the classes provided by the overloaded records machinery supply sufficient information (e.g. `HasField` gives us an accessor, but doesn't say anything about position). So then the only "easy" way right now seems to define specific `HasField` instances for the wrapped record, using TH.

I'll try to think about this some more.

**Comment by adamgundry** [comment]
Interesting question. I agree with @kosmikus that `HasField` probably doesn't provide enough information, in its current draft form.

This did inspire me to have a go at implementing a generic version of `HasField` using the new type-level metadata, which turns out to be possible but painful: https://gist.github.com/adamgundry/2eea6ca04fd6e5b6e76ce9bfee454a6b

**Comment by kosmikus** [comment]
Somewhat related efforts: https://github.com/kcsongor/generic-records and https://github.com/kosmikus/records-sop, the latter based on experimental support for type-level metadata in generics-sop.

**Comment by danidiaz** [comment]
One small additional comment about this issue: I've just realised that record pattern synonyms can come in handy when constructing sum-of-products representations. For example:

    {-# language DeriveGeneric #-}
    {-# language TypeFamilies #-}
    {-# language TypeOperators #-}
    {-# language PatternSynonyms #-}

    import Data.Text
    import qualified GHC.Generics as GHC
    import Generics.SOP
    import Text.Read

    data Foo = Foo { x :: Int, y :: Text } deriving (Show, GHC.Generic)

    instance Generic Foo

    pattern Foo' :: t Int -> t Text -> SOP t (Code Foo)
    pattern Foo' {x', y'} = SOP (Z (x' :* y' :* Nil))

    readFooMaybe :: SOP (IO :.: Maybe) (Code Foo)
    readFooMaybe = Foo'
                 {
                    x' = Comp (fmap readMaybe getLine)
                 ,  y' = Comp (fmap readMaybe getLine)
                 }

    main :: IO ()
    main = do
       r <- hsequence' readFooMaybe
       print r

**Comment by kosmikus** [comment]
Yes, true. It would sometimes be nice to have more convenient ways of creating either SOPs or POPs compatible with a particular datatype directly. I don't currently see a way to do that generically though. Introducing new top-level declarations of any kind is still a feature that only TH has.

Your particular example could better be written as just
```
readFooMaybe = SOP (Z (hcpure (Proxy :: Proxy Read) (Comp (fmap readMaybe getLine))))
```
but I guess you know that and you just simplified it for this example.

**Comment by kosmikus** [comment]
@danidiaz Just to let you know, with generics-sop with type-level metadata and also records-sop now released, I'm actually looking at this again. Also because it's relevant for my slowly ongoing work on sop-based generic data migrations.
